luarules/gadgets/unit_dgun_behaviour.lua

```diff
@@ -57,63 +82,119 @@ local function addVolumetricDamage(projectileID)
 		ignoreOwner = dgunWeapons[weaponDefID].noSelfDamage,
 		damageGround = true,
 	}
-	Spring.SpawnExplosion(x, y ,z, 0, 0, 0, explosionParame)
+
+	spSpawnExplosion(x, y ,z, 0, 0, 0, explosionParame)
 end
 
 function gadget:ProjectileCreated(proID, proOwnerID, weaponDefID)
 	if dgunWeapons[weaponDefID] then
 		flyingDGuns[proID] = true
+		dgunTimeouts[proID] = (frameCounter + dgunWeaponsTTL[weaponDefID])
+
+		local posX, posY, posZ = spGetProjectilePosition(proID)
+		dgunOrigins[proID] = {posX, posY, posZ}
 	end
 end
 
 function gadget:ProjectileDestroyed(proID)
 	flyingDGuns[proID] = nil
 	groundedDGuns[proID] = nil
+	dgunTimeouts[proID] = nil
 end
 
-function gadget:GameFrame()
+function gadget:GameFrame(frame)
 	for proID in pairs(flyingDGuns) do
 		-- Fireball is hitscan while in flight, engine only applies AoE damage after hitting the ground,
 		-- so we need to add the AoE damage manually for flying projectiles
 		addVolumetricDamage(proID)
 
-		local x, y, z = Spring.GetProjectilePosition(proID)
-		local h = Spring.GetGroundHeight(x, z)
+		local x, y, z = spGetProjectilePosition(proID)
+		local h = spGetGroundHeight(x, z)
 
 		if y < h + 1 or y < 0 then -- assume ground or water collision
 			-- normalize horizontal velocity
-			local dx, _, dz, speed = Spring.GetProjectileVelocity(proID)
+			local dx, _, dz, speed = spGetProjectileVelocity(proID)
 			local norm = speed / math.sqrt(dx^2 + dz^2)
 			local ndx = dx * norm
 			local ndz = dz * norm
-			Spring.SetProjectileVelocity(proID, ndx, 0, ndz)
+			spSetProjectileVelocity(proID, ndx, 0, ndz)
 
 			groundedDGuns[proID] = true
 			flyingDGuns[proID] = nil
 		end
 	end
 
 	for proID in pairs(groundedDGuns) do
-		local x, y, z = Spring.GetProjectilePosition(proID)
+		local x, y, z = spGetProjectilePosition(proID)
 		-- place projectile slightly under ground to ensure fiery trail
 		local verticalOffset = 1
-		Spring.SetProjectilePosition(proID, x, math.max(Spring.GetGroundHeight(x, z), 0) - verticalOffset, z)
+		spSetProjectilePosition(proID, x, math.max(spGetGroundHeight(x, z), 0) - verticalOffset, z)
 
 		-- NB: no removal; do this every frame so that it doesn't fly off a cliff or something
 	end
+
+	-- Without defining a time to live (TTL) for the DGun, it will live forever until it reaches maximum range. This means it would deal infinite damage to shields until it depleted them.
+	if next(dgunTimeouts) == nil then
+		frameCounter = 0
```

Comment:
> I had issues where if I counted frames it would loop back in on itself and cause sometimes for the D gun to live forever and sometimes to expire immediately.
> 
That sounds like integer overflow, but the number of frames elapsed in a game should never come close to approaching the value needed for that to happen.

> I created my own localized frame counter. But that created the issue that it would be a infinitely expanding number that would take up more and more memory.
> 
An integer variable will take up the same amount of memory, regardless of the value stored inside it. That amount of memory is also negligible - 8 bytes, you could store 125,000 of them in 1MB of memory.